File: problems/336_palindrome_pairs/solution.py
class Solution:
    def palindromePairs(self, words: list[str]) -> list[list[int]]:
        word_map = {w: i for i, w in enumerate(words)}
        res = []
        for i, word in enumerate(words):
            for j in range(len(word) + 1):
                prefix = word[:j]
                suffix = word[j:]
                if prefix == prefix[::-1]:
                    rev_suffix = suffix[::-1]
                    if rev_suffix in word_map and word_map[rev_suffix] != i:
                        res.append([word_map[rev_suffix], i])
                if j != len(word) and suffix == suffix[::-1]:
                    rev_prefix = prefix[::-1]
                    if rev_prefix in word_map and word_map[rev_prefix] != i:
                        res.append([i, word_map[rev_prefix]])
        return res

File: problems/336_palindrome_pairs/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    vector<vector<int>> palindromePairs(vector<string>& words) {
        unordered_map<string, int> map;
        for (int i = 0; i < words.size(); i++) map[words[i]] = i;
        vector<vector<int>> res;
        for (int i = 0; i < words.size(); i++) {
            for (int j = 0; j <= (int)words[i].size(); j++) {
                string prefix = words[i].substr(0, j);
                string suffix = words[i].substr(j);
                if (isPalin(prefix)) {
                    string rev = suffix;
                    reverse(rev.begin(), rev.end());
                    if (map.count(rev) && map[rev] != i)
                        res.push_back({map[rev], i});
                }
                if (j != (int)words[i].size() && isPalin(suffix)) {
                    string rev = prefix;
                    reverse(rev.begin(), rev.end());
                    if (map.count(rev) && map[rev] != i)
                        res.push_back({i, map[rev]});
                }
            }
        }
        return res;
    }
    bool isPalin(string& s) {
        int l = 0, r = s.size() - 1;
        while (l < r) if (s[l++] != s[r--]) return false;
        return true;
    }
};
